monai/apps/pathology/datasets.py

```diff
@@ -0,0 +1,146 @@
+        image_reader_name: str = "cuCIM",
+        transform: Union[Sequence[Callable], Callable] = None,
+    ):
+        if type(region_size) == int:
+            self.region_size = (region_size, region_size)
+        else:
+            self.region_size = region_size
+
+        if type(grid_shape) == int:
+            self.grid_shape = (grid_shape, grid_shape)
+        else:
+            self.grid_shape = grid_shape
+
+        self.patch_size = patch_size
+        self.sub_region_size = (self.region_size[0] / self.grid_shape[0], self.region_size[1] / self.grid_shape[1])
+
+        self.transform = transform
+        self.samples = data
+        self.image_path_list = list({x["image"] for x in self.samples})
+
+        self.image_reader_name = image_reader_name
+        self.image_reader = WSIReader(image_reader_name)
+        self.wsi_object_dict = None
+        if self.image_reader_name != "openslide":
+            # OpenSlide causes memeory issue if we prefetch image objects
+            self._fetch_wsi_objects()
+
+    def _fetch_wsi_objects(self):
+        self.wsi_object_dict = {}
+        for image_path in self.image_path_list:
+            self.wsi_object_dict[image_path] = self.image_reader.read(image_path)
+
+    def __len__(self):
+        return len(self.samples)
+
+    def __getitem__(self, index):
+        sample = self.samples[index]
+        if self.image_reader_name == "openslide":
+            img_obj = self.image_reader.read(sample["image"])
+        else:
+            img_obj = self.wsi_object_dict[sample["image"]]
+        location = [sample["location"][i] - self.region_size[i] // 2 for i in range(len(self.region_size))]
+        images, _ = self.image_reader.get_data(
+            img=img_obj,
+            location=location,
+            size=self.region_size,
+            grid_shape=self.grid_shape,
+            patch_size=self.patch_size,
+        )
+        labels = np.array(sample["label"], dtype=np.float32)[:, np.newaxis, np.newaxis]
+        patches = [{"image": images[i], "label": labels[i]} for i in range(len(sample["label"]))]
+        if self.transform:
+            patches = self.transform(patches)
+        return patches
+
+
+class SmartCachePatchWSIDataset(SmartCacheDataset):
+    """
+    Add SmartCache functionality to PatchWSIDataset
+    """
+
+    def __init__(
+        self,
+        data: List,
+        region_size: Union[int, Tuple[int, int]],
+        grid_shape: Union[int, Tuple[int, int]],
+        patch_size: int,
+        image_reader_name: str = "cuCIM",
+        transform: Union[Sequence[Callable], Callable] = None,
+        replace_rate: float = 0.5,
+        cache_num: int = sys.maxsize,
+        cache_rate: float = 1.0,
+        num_init_workers: Optional[int] = None,
+        num_replace_workers: Optional[int] = None,
+    ):
+        extractor = PatchWSIDataset(data, region_size, grid_shape, patch_size, image_reader_name)
+        super().__init__(
+            data=extractor,
```

Comment:
Hi @Nic-Ma, the `data` argument in `SmartCache` supposed to be 'Sequence` but I am passing a `Dataset` instance. Do you know how should I fix this?

Reply:
As we may add SmartCache wrapper later, I think you can ignore this error so far, refer to:
https://github.com/Project-MONAI/MONAI/blob/master/monai/data/png_writer.py#L80

Thanks.